src/pipeline/src/etl.rs

```diff
@@ -97,13 +220,110 @@ impl<T> Pipeline<T>
 where
     T: Transformer,
 {
-    pub fn exec(&self, val: value::Value) -> Result<T::Output, String> {
-        let mut val = val;
+    fn exec_map<'a>(&self, map: &'a mut Map) -> Result<&'a mut Map, String> {
+        let mut v = map;
         for processor in self.processors.iter() {
-            val = processor.exec(val)?;
+            v = processor.exec_map(v)?;
         }
+        Ok(v)
+    }
+
+    pub fn exec(&self, mut val: value::Value) -> Result<T::Output, String> {
+        let result = match val {
+            value::Value::Map(ref mut map) => {
+                self.exec_map(map)?;
+                val
+            }
+            value::Value::Array(arr) => arr
+                .values
+                .into_iter()
+                .map(|mut v| match v {
+                    value::Value::Map(ref mut map) => {
+                        self.exec_map(map)?;
+                        Ok(v)
+                    }
+                    _ => Err(format!("expected a map, but got {}", v)),
+                })
+                .collect::<Result<Vec<value::Value>, String>>()
+                .map(|values| value::Value::Array(value::Array { values }))?,
+            _ => return Err(format!("expected a map or array, but got {}", val)),
+        };
+
+        self.transformer.transform(result)
+    }
+
+    pub fn exec_mut(&self, val: &mut Vec<value::Value>) -> Result<T::VecOutput, String> {
+        for processor in self.processors.iter() {
+            processor.exec_mut(val)?;
+        }
+
+        self.transformer.transform_mut(val)
+    }
+
+    pub fn preprepase(&self, val: serde_json::Value, result: &mut [Value]) -> Result<(), String> {
```

Comment:
Too much `pre`? Is `prepare` enough?

Reply:
Add some unittests and doc comments to let everyone see what the expected behavior is?

Reply:
Add a simple test to show how it works.